### DISCUSSION TITLE
Is there a way (without eval) to access object.__setattr__

### DISCUSSION THREAD

**Discussion by samuelcolvin** [PyO3/pyo3#2321] [Questions]
I use `object.__setattr__` to set attributes while bypassing a custom `__setattr__` method, see [here](https://github.com/samuelcolvin/pydantic/blob/8997cc5961139dd2695761a33c06a66adbf1430a/pydantic/main.py#L297).

My Current logic for doing this in rust is:

```rs

static SET_ATTR_CELL: GILOnceCell<PyObject> = GILOnceCell::new();

pub fn get_setattr(py: Python<'_>) -> &PyAny {
    SET_ATTR_CELL
        .get_or_init(py, || py.eval("object.__setattr__", None, None).unwrap().into())
        .as_ref(py)
}

        // Then later
        let setattr = get_setattr(py);
        setattr.call((&instance, intern!(py, "foobar"), foobar), None)?;
```

I'm wondering if there is a cleaner and more performant way of getting access to `object.__setattr__` or a similar method?

The [`setattr`](https://docs.rs/pyo3/latest/pyo3/prelude/struct.Py.html#method.setattr) method doesn't work as it doesn't bypass the setattr method on the class/instance in question.

**Comment by samuelcolvin** [comment]
I've found a work around, what I needed was [`PyObject_GenericSetAttr`](https://docs.python.org/3/c-api/object.html#c.PyObject_GenericSetAttr), then I implemented my own `force_setattr`

```rs

pub fn force_setattr<N, V>(obj: &PyObject, py: Python<'_>, attr_name: N, value: V) -> PyResult<()>
where
    N: ToPyObject,
    V: ToPyObject,
{
    attr_name.with_borrowed_ptr(py, move |attr_name| {
        value.with_borrowed_ptr(py, |value| unsafe {
            error_on_minusone(py, ffi::PyObject_GenericSetAttr(obj.as_ptr(), attr_name, value))
        })
    })
}

// Defined here as it's not exported by pyo3
#[inline]
fn error_on_minusone(py: Python<'_>, result: c_int) -> PyResult<()> {
    if result != -1 {
        Ok(())
    } else {
        Err(PyErr::fetch(py))
    }
}
```

**Reply by samuelcolvin** [reply]
Still, I think it might be useful for others if this method was include in pyo3?

**Reply by mejrs** [reply]
Possibly -can you expand about your use case for this?

**Reply by samuelcolvin** [reply]
As per the link above, I'm creating pydantic class instances in rust, but pydantic classes have a `__setattr__` method to allow valuation on assignment. but when setting `__dict__` and `__fields_set__` I need to bypass that method.

So basically the rust equivalent of  [this](https://github.com/samuelcolvin/pydantic/blob/8997cc5961139dd2695761a33c06a66adbf1430a/pydantic/main.py#L402-L403).

Sorry the code isn't open source yet, should be in the coming days - I'll comment here when it is.

**Reply by samuelcolvin** [reply]
[this](https://github.com/samuelcolvin/pydantic-core/blob/459e4b406e5dcc7393c6aa2391e2b5ce616a183d/src/validators/model_create.rs#L65-L101) is the code in question.

**Reply by mejrs** [reply]
I'm weakly against adding an api for this. This use case seems pretty niche. And I fear it would be easily confused with `setattr` 😐 .

**Reply by samuelcolvin** [reply]
That's fair I think.

**Comment by mejrs** [comment]
You can get it from `object` itself:

```rust
let setattr = PyModule::import(py, "builtins")?
            .getattr("object")?
            .getattr("__setattr__")?;
```
